flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java

```diff
@@ -101,32 +97,57 @@ public void collect(StreamRecord<OUT> record) {
     }
 
     private <X> void pushToRecordWriter(StreamRecord<X> record) {
-        serializationDelegate.setInstance(record);
+        // StreamStatus.IDLE requires that no records nor watermarks travel through the branch
+        // in order to keep the older behaviour that records could've been generated down the
+        // pipeline even though the sources were idle we go through a short ACTIVE/IDLE loop
+        if (announcedStatus.isIdle()) {
+            writeStreamStatus(StreamStatus.ACTIVE);
+        }
 
+        serializationDelegate.setInstance(record);
         try {
             recordWriter.emit(serializationDelegate);
         } catch (Exception e) {
             throw new RuntimeException(e.getMessage(), e);
         }
+
+        if (announcedStatus.isIdle()) {
+            writeStreamStatus(StreamStatus.IDLE);
+        }
```

Comment:
Can we factor out some reusable pattern?

```
try (announcedStatus.ensureActive(this::writeStreamStatus)) {
        serializationDelegate.setInstance(record);
        try {
            recordWriter.emit(serializationDelegate);
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        }
}
```

Reply:
Nice one! Will do.